NIFI-15932: Implemented ability to migrate a Versioned flow's Assets … (#11240)
apache/nifi: nifi-docs/src/main/asciidoc/developer-guide.adoc

`mvn clean install -Pcontrib-check`

== Supporting migration from a Versioned Process Group

> Connectors can opt in to a one-time migration flow that uses a source Versioned Process Group as a reference for populating a freshly created Connector.

Reviewer: ```suggestion
Connectors can implement a one-time migration flow that uses a source Versioned Process Group as a reference for populating a freshly created Connector.
```

Author: [Sonnet 5] Updated as suggested.

The two phases together form a single, all-or-nothing migration. The framework only commits the merged configuration onto the Connector's persisted active configuration after both `migrateState` and the framework's subsequent state-application step have succeeded. Any failure in either phase rolls the migration back: the managed Process Group is restored from `getInitialFlow()`, the working configuration copy is discarded, any assets copied via `copyAssetFromSource(...)` are deleted, and the Connector's persisted configuration remains unchanged.

=== Accessing the source flow

`ConnectorMigrationContext` exposes the source `VersionedExternalFlow` using `getSourceFlow()`.
The source includes:

- The exported `VersionedProcessGroup`
- Parameter contexts
- Referenced assets
- Attached `VersionedComponentState` for `@Stateful` components
- Export metadata such as flow name, registry identifiers, and version

The same context also indicates whether the request came from a local Process Group on the same NiFi instance or from an uploaded payload using `isLocalMigration()`.

> The active `FlowContext` is available via `context.getActiveFlowContext()` and is read-only during migration: it forwards reads (configuration, managed Process Group) to the live active flow context but rejects `updateFlow(...)` with `IllegalStateException`. The Connector should treat this object as a window onto the current state of the Connector, not as a write target.

Reviewer: It is mentioned later that the applyUpdate will be performed to update the active context but it might be worth mentioning here as well.

Author: [Sonnet 5] Added a mention of applyUpdate to this section as well.

`setComponentState` only stages the desired state; the framework writes it after `migrateState` returns. If a Connector calls `setComponentState` more than once for the same versioned identifier, only the last call is applied.

=== Rewriting parameters and assets

The exported source flow contains parameter contexts and any asset references that were attached to exported parameters.
Connectors are responsible for mapping those values into their own target shape inside `migrateConfiguration`.

A Connector can copy an asset from a local source Process Group into its own asset namespace using `ConnectorMigrationContext.copyAssetFromSource(String)`.
This method is available only for local migrations.
It throws `IllegalArgumentException` when the supplied source asset identifier is null or blank, and `IllegalStateException` when invoked for an uploaded-payload migration.
Uploaded payloads do not have access to a live source asset namespace, so asset references from uploaded payloads must be handled outside of the migration request.

> 

Reviewer: Code snippet for Asset Migration might be helpful.  As asset needs to be migrated and reference set.

```
    final ConnectorMigrationContext context = ...
    final VersionedParameter sourceParameter = ...
    final String targetStepName = ...
    final String targetPropertyName = ...
    if (sourceParameter == null || sourceParameter.getReferencedAssets() == null
            || sourceParameter.getReferencedAssets().isEmpty() || !context.isLocalMigration()) {
        return;
    }
    final Set<String> copiedAssetIds = new LinkedHashSet<>();
    for (final VersionedAsset asset : sourceParameter.getReferencedAssets()) {
        if (asset.getIdentifier() != null && !asset.getIdentifier().isBlank()) {
            copiedAssetIds.addAll(context.copyAssetFromSource(asset.getIdentifier()).getAssetIdentifiers());
        }
    }
    if (!copiedAssetIds.isEmpty()) {
        context.setValueReference(targetStepName, targetPropertyName, new AssetReference(copiedAssetIds));
    }
```

Author: [Sonnet 5] Added a code example for asset migration to the developer guide, based on the pattern used by MigrationTargetConnector in the system test extensions.

- `isMigrationSupported(...)` must not mutate flow state, parameter values, or assets.
- `migrateConfiguration` and `migrateState` should either complete successfully or throw an exception. Any thrown exception aborts the migration and rolls back every side effect: copied assets are deleted, the managed Process Group is restored from `getInitialFlow()`, every component's `StateManager` state is cleared, and the persisted active configuration is left unchanged.
- Migration is rejected when the Connector's currently managed flow no longer matches the flow returned by its `getInitialFlow()`.
  The framework compares the two structurally (ignoring environmental differences such as component scheduled state) at the time `migrateConfiguration` is invoked, and again whenever the `MIGRATE` allowable action is evaluated, so the user is shown why migration is unavailable.
  This check is restart-safe because both inputs are recomputed each time from durable sources: the Connector's persisted configuration drives `getInitialFlow()`, and the current managed flow is read from the live flow tree restored from `flow.json.gz`.
- Migration is one-shot from the framework's point of view.

>   Once the Connector's flow has been updated by a migration, its currently managed flow no longer matches `getInitialFlow()`, and a subsequent migration request will be rejected until the Connector is discarded and a fresh one is created.

Reviewer: Is the point about getInitialFlow still true with the recent changes?

Author: [Sonnet 5] No — this text was stale relative to the eligibility check added in the July 16 commit, which now uses Connector.isModified() (configuration deviates from defaults, or any managed component has stored state) rather than a structural comparison against getInitialFlow(). Reworded this section to describe the current check.
